New Module: Add github_deploy_key module (source_control/github_deploy_key)

bincyber opened the pull request:
##### SUMMARY

Added a new module for managing deploy keys for GitHub repositories.

Currently, this can be accomplished using the uri module, however it requires some effort to achieve idempotency. It's better to have a specific module that can do the same thing in 1 task.

Additionally, this module supports authenticating to GitHub's API using username and password, username and password and 2-factor authentication code (OTP), OAuth2 token, or personal access token.

##### ISSUE TYPE
 - New Module Pull Request


##### COMPONENT NAME

github_deploy_key

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /vagrant/ansible.cfg
  configured module search path = [u'modules/']
  python version = 2.7.5 (default, Nov 20 2015, 02:00:19) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```


##### ADDITIONAL INFORMATION

A module for managing deploy keys for GitHub repositories was initially requested in ansible/ansible-modules-core#2626. I began a discussion on the [ansible-devel](https://groups.google.com/forum/#!msg/ansible-devel/XY9XUYFRZvQ/yqOhynX1EQAJ;context-place=forum/ansible-devel) Google Group, but nobody was interested in developing this module. So, I finally got around to doing it myself.

bincyber committed:
Added new module: github_deploy_key

added a new module for managing deploy keys for GitHub repositories

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
```
Add GNU GPL like [github_issue](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/source_control/github_issue.py#L4-L17)

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+
+class GithubDeployKey(object):
+    def __init__(self, module=None, url=None, state=None, username=None, password=None, token=None, otp=None):
+        self.module     = module
```
Remove spaces before operator `=`

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+        if self.otp is not None:
+            self.headers = {"X-GitHub-OTP": self.otp}
+            self.auth = (self.username, self.password)
+        elif username is not None and password is not None:
```
This condition can come early

```
if username is not None and password is not None:
    self.auth = (self.username, self.password)
    if self.otp is not None:
        self.headers = {"X-GitHub-OTP": self.otp}
```

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+        elif username is not None and password is not None:
+            self.auth = (self.username, self.password)
+        else:
+            self.headers = {"Authorization": "token " + self.token}
```
I would prefer `format` here.

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+            self.module.fail_json(msg="Failed to connect to github.com due to connection timeout")
+
+        if resp.status_code == 200:
+            if len(resp.json()) > 0:
```
This can be simplified as `if resp.json()`

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+        try:
+            resp = requests.get(self.url, auth=self.auth, headers=self.headers, timeout=self.timeout)
+        except:
+            self.module.fail_json(msg="Failed to connect to github.com due to connection timeout")
```
Print exception using `get_exception()`, it will easy for debugging purpose.

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            owner       = dict(required=True, type='str', aliases=['account', 'organization']),
```
Remove spaces before operator `=`

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+            supports_check_mode = False,
+        )
+
+    owner       = module.params['owner']
```
Remove spaces before operator.

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+
+    def get_existing_key(self, key, title, force):
+        try:
+            resp = requests.get(self.url, auth=self.auth, headers=self.headers, timeout=self.timeout)
```
use `fetch_url` instead of `requests.get`

gundalow commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+DOCUMENTATION = '''
+---
+module: github_deploy_key
+version_added: "2.3"
```
version_added: "2.4"

gundalow commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+            force       = dict(required=False, type='bool', default=False),
+            username    = dict(required=False, type='str'),
+            password    = dict(required=False, type='str', no_log=True),
+            otp         = dict(required=False, type='int', aliases=['2fa_token']),
```
Please add `no_log=True`

gundalow commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+            username    = dict(required=False, type='str'),
+            password    = dict(required=False, type='str', no_log=True),
+            otp         = dict(required=False, type='int', aliases=['2fa_token']),
+            token       = dict(required=False, type='str')
```
Please add `no_log=True`

gundalow commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -0,0 +1,299 @@
+            required_one_of = [
+                ['username', 'token']
+            ],
+            supports_check_mode = False,
```
What would be needed to support checkmode?h

gundalow commented:
@pcgentry @adrianmoisey I believe you two are maintainers for other GitHub modules, it would be good to get your input on this PR. Thanks in advance.

bincyber committed:
Updated github_deploy_key module based on feedback

- added GPL header
- switched to using fetch_url instead of requests
- changed version to 2.4
- set no_log for otp and token arguments
- implemented check mode

bincyber commented:
Thanks @Akasurde and @gundalow for your feedback on this PR. I've addresses all of your comments in my latest commit.

Akasurde commented on lib/ansible/modules/source_control/github_deploy_key.py:
```diff
@@ -11,11 +25,11 @@
 short_description: Manages deploy keys for GitHub repositories.
 description:
-  - Adds or removes deploy keys for GitHub repositories. Supports authentication using username and password, username and password and 2-factor authentication code (OTP), OAuth2 token, or personal access token. Refer to GitHub's API documentation here: U(https://developer.github.com/v3/repos/keys/).
+  - "Adds or removes deploy keys for GitHub repositories. Supports authentication using username and password, username and password and 2-factor authentication code (OTP), OAuth2 token, or personal access token. Refer to GitHub's API documentation here: U(https://developer.github.com/v3/repos/keys/)."
```
Line too long.

Akasurde commented:
@bincyber Shippable is not happy with changes. Can you please fix [issues](https://app.shippable.com/github/ansible/ansible/runs/21883/1/tests) to match coding style?

bincyber committed:
made github_deploy_key module PEP8 compliant

bincyber commented:
@Akasurde I've corrected the PEP8 issues.

Akasurde reviewed (approved):
LGTM

gundalow commented:
Looking good, would like to get it review by some other people

@pcgentry @adrianmoisey I believe you two are maintainers for other GitHub modules, it would be good to get your input on this PR. Thanks in advance.

gundalow commented:
@bincyber Poke me in a week if and if there isn't any other feedback I'll merge it

bincyber commented:
Thanks @Akasurde @gundalow !

bincyber commented:
@gundalow Can we get this PR merged please?

gundalow commented:
Triggering ci

gundalow commented:
Merged into `devel`, so this will be released in Ansible 2.4

bincyber commented:
Thanks @gundalow, appreciate it
